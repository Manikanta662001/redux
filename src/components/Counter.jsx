import React, { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import {
  decrement,
  increment,
  increment_by_async,
  increment_by_value,
  reset,
} from "../store/actions/actions";
function Counter() {
  const countVal = useSelector((state) => state.counterReducer.countval);
  const dispatch = useDispatch();
  const [val, setVal] = useState(0);
  return (
    <div>
      <h4>{countVal}</h4>
      <div>
        <button onClick={() => dispatch(increment())}>+</button>
        <button onClick={() => dispatch(decrement())}>-</button>
        <button onClick={() => dispatch(reset())}>RESET</button>
        <button
          onClick={() => {
            setTimeout(() => {
              dispatch(increment_by_async());
            }, 1500);
          }}
        >
          + by some delay
        </button>
      </div>
      <input
        type="number"
        placeholder="Enter A Number"
        value={val}
        onChange={(e) => setVal(e.target.value)}
      />
      <button onClick={()=>dispatch(increment_by_value(Number(val)))}>increment by value</button>
    </div>
  );
}

export default Counter;
