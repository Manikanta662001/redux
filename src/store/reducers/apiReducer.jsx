import {
  FETCH_USERS_FAILURE,
  FETCH_USERS_REQUEST,
  FETCH_USERS_SUCCESS,
} from "../actions/actionTypes";

const initialState = {
  users: [],
  isLoading: false,
  errormsg: "",
};
const apiReducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_USERS_REQUEST:
      return { ...state, isLoading: true };
    case FETCH_USERS_SUCCESS:
      return { ...state, isLoading: false, users: action.payload ,errormsg:""};
    case FETCH_USERS_FAILURE:
      return { ...state, isLoading: false, errormsg: action.payload ,users:[]};
    default:
      return state;
  }
};

export default apiReducer;
