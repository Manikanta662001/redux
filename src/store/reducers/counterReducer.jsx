import { DECREMENT, INCREMENT, INCREMENT_BY_VALUE, RESET } from "../actions/actionTypes";

const initialState = { countval: 0 };
const counterReducer = (state = initialState, action) => {
  switch (action.type) {
    case INCREMENT:
      return { ...state, countval: state.countval + 1 };
    case DECREMENT:
      return { ...state, countval: state.countval - 1 };
    case RESET:
      return { ...state, countval: state.countval - state.countval };
    case INCREMENT_BY_VALUE:
      return { ...state, countval: state.countval + action.payload };
    default:
        return state
  }
};

export default counterReducer;
