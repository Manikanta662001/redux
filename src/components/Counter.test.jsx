import { render, screen, cleanup, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Counter from "./Counter";
import { combineReducers, configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import counterReducer from "../store/reducers/counterReducer";
import apiReducer from "../store/reducers/apiReducer";
const rootReducer = combineReducers({
  counterReducer,
  apiReducer,
});


const renderWithProvider = (component, { initialState } = {}) => {
  const store = configureStore({
    reducer: rootReducer,
    preloadedState: initialState,
  });
  return { ...render(<Provider store={store}>{component}</Provider>), store };
};
describe("Counter Component", () => {
  const initialState = {
    counterReducer: { countval: 0 },
    apiReducer: { users: [], isLoading: false, errormsg: "" },
  };
  const user = userEvent.setup();
  test("renders correctly and Match the Snapshot", () => {
    const { asFragment } = renderWithProvider(<Counter />, { initialState });
    expect(asFragment(<Counter />)).toMatchSnapshot();
    expect(screen.getByRole("heading", { level: 4 })).toHaveTextContent("0");
  });
  afterEach(cleanup);

  test("dispatches increment action when we click + button", async () => {
    const { store } = renderWithProvider(<Counter />, { initialState });
    const incButton = screen.getByRole("button", { name: "+" });
    await user.click(incButton);
    expect(screen.getByRole("heading", { level: 4 })).toHaveTextContent("1");
    //to test the store value
    expect(store.getState().counterReducer.countval).toBe(1);
  });
  test("dispatches decrement action when we click - button", async () => {
    const { store } = renderWithProvider(<Counter />, { initialState });
    const decButton = screen.getByRole("button", { name: "-" });
    await user.click(decButton);
    expect(screen.getByRole("heading", { level: 4 })).toHaveTextContent("-1");
    //to test the store value
    expect(store.getState().counterReducer.countval).toBe(-1);
  });
  test("dispatches reset action when we click RESET button", async () => {
    const { store } = renderWithProvider(<Counter />, { initialState });
    const resetButton = screen.getByRole("button", { name: "RESET" });
    await user.click(resetButton);
    expect(screen.getByRole("heading", { level: 4 })).toHaveTextContent("0");
    //to test the store value
    expect(store.getState().counterReducer.countval).toBe(0);
  });
  test("increments by some delay", async () => {
    const { store } = renderWithProvider(<Counter />, { initialState });
    const delayButton = screen.getByRole("button", { name: "+ by some delay" });
    await user.click(delayButton);
    await waitFor(
      () => {
        expect(store.getState().counterReducer.countval).toBe(1);
        expect(screen.getByRole("heading", { level: 4 })).toHaveTextContent(
          "1"
        );
      },
      { timeout: 2000 }
    );
  });
  test("increment by value", async () => {
    const { store } = renderWithProvider(<Counter />, { initialState });
    const input = screen.getByRole("spinbutton");
    await user.type(input, "3");
    const valueButton = screen.getByRole("button", {
      name: "increment by value",
    });
    await user.click(valueButton);
    expect(screen.getByRole("heading", { level: 4 })).toHaveTextContent("3");
  });
});
