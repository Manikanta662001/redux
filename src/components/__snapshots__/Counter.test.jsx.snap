// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component renders correctly and Match the Snapshot 1`] = `
<DocumentFragment>
  <div>
    <h4>
      0
    </h4>
    <div>
      <button>
        +
      </button>
      <button>
        -
      </button>
      <button>
        RESET
      </button>
      <button>
        + by some delay
      </button>
    </div>
    <input
      placeholder="Enter A Number"
      type="number"
      value="0"
    />
    <button>
      increment by value
    </button>
  </div>
</DocumentFragment>
`;
